cmd/deployment-check/deployment.go

```diff
@@ -0,0 +1,704 @@
+// be an old deployment belonging to this check that should be deleted.
+func findPreviousDeployment() (bool, error) {
+
+	log.Infoln("Attempting to find previously created deployment(s) belonging to this check.")
+
+	deploymentList, err := client.AppsV1().Deployments(checkNamespace).List(metav1.ListOptions{})
+	if err != nil {
+		log.Infoln("error listing deployments:", err)
+		return false, err
+	}
+	log.Debugln("Found", len(deploymentList.Items), "deployment(s)")
+
+	if debug { // Print out all the found deployments if debug logging is enabled.
+		for _, deployment := range deploymentList.Items {
+			log.Debugln(deployment.Name)
+		}
+	}
+
+	// Iterate through deployments and look for previous deployments.
+	for _, deployment := range deploymentList.Items {
+
+		// Check using names.
+		if deployment.Name == checkDeploymentName {
+			log.Infoln("Found an old deployment belonging to this check:", deployment.Name)
+			return true, nil
+		}
+
+		// check using labels
+		// labels := deployment.Labels
+		// for k, v := range labels {
+		// 	if k == defaultLabelKey && v != defaultLabelValueBase+strconv.Itoa(int(now.Unix())) {
+		// 		log.Infoln("Found an old deployment belonging to this check.")
+		// 		return true, nil
+		// 	}
+		// }
+	}
+
+	log.Infoln("Did not find any old deployment(s) belonging to this check.")
+	return false, nil
+}
+
+// updateDeployment performs an update on a deployment with a given deployment configuration.  The DeploymentResult
+// channel is notified when the rolling update is complete.
+func updateDeployment(deploymentConfig *v1.Deployment) chan DeploymentResult {
+
+	updateChan := make(chan DeploymentResult)
+
+	go func() {
+		log.Infoln("Performing rolling-update on deployment", deploymentConfig.Name, "to ["+deploymentConfig.Spec.Template.Spec.Containers[0].Image+"]")
+
+		defer close(updateChan)
+
+		result := DeploymentResult{}
+
+		// Get the names of the current pods and ignore them when checking for completed rolling-update.
+		log.Infoln("Creating a blacklist with the current pods that exist.")
+		oldPodNames := getPodNames()
+
+		deployment, err := client.AppsV1().Deployments(checkNamespace).Update(deploymentConfig)
+		if err != nil {
+			log.Infoln("Failed to update deployment in the cluster:", err)
+			result.Err = err
+			updateChan <- result
+			return
+		}
+
+		// Watch that it is up.
+		watch, err := client.AppsV1().Deployments(checkNamespace).Watch(metav1.ListOptions{
+			Watch:         true,
+			FieldSelector: "metadata.name=" + deployment.Name,
+			// LabelSelector: defaultLabelKey + "=" + defaultLabelValueBase + strconv.Itoa(int(now.Unix())),
+		})
+		if err != nil {
+			result.Err = err
+			updateChan <- result
+			return
+		}
+
+		// Stop the watch on each loop because we will create a new one.
+		defer watch.Stop()
+
+		log.Debugln("Watching for deployment rolling-update to complete.")
+		newPodStatuses := make(map[string]bool)
```

Comment:
consider a refactor that that goes off of the status on a deployment instead of a watch with your own state in a map.  This way the deployment logic is exactly aligned with your determination of deployment success.

![image](https://user-images.githubusercontent.com/98695/67250449-49744c00-f420-11e9-8e33-39c9d02cf73c.png)